# Chapter 9: Backpropagation

Start with a simplified forward pass with just one neuron.

Let’s backpropagate the ReLU function for a single neuron, then intend to minimize the output for this single neuron as a practice to show how we can leverage the chain rule with derivatives and partial derivatives.

We will start by <b>minimizing this more basic output</b> before jumping to the full network and overall loss.

In [1]:
x = [1.0, -2.0, 3.0] # input values
w = [-3.0, -1.0, 2.0] # weights
b = 1.0 # bias

xw0 = x[0] * w[0]
print(xw0)

-3.0


<center><img src='./image/9-1.png' style='width: 70%'/><font color='gray'><i>The first input and weight multiplication.</i></font></center>

In [2]:
xw1 = x[1] * w[1]
xw2 = x[2] * w[2]

print(xw0, xw1, xw2)


-3.0 2.0 6.0


<center><img src='./image/9-2.png' style='width: 70%'/><font color='gray'><i>Input and weight multiplication of all of the inputs.</i></font></center>

Perform a sum of all weighted inputs with a bias.

In [3]:
# Adding weighted inputs and a bias
z = xw0 + xw1 + xw2 + b
print(z)

6.0


<center><img src='./image/9-3.png' style='width: 70%'/><font color='gray'><i>Weighted inputs and bias addition.</i></font></center>

In [4]:
# ReLU activation function
y = max(z, 0)
print(y)

6.0


<center><img src='./image/9-4.png' style='width: 70%'/><font color='gray'><i>ReLU activation applied to the neuron output.</i></font></center>

This is the full forward pass through a single neuron and a ReLU activation function.

Let’s treat all of these chained functions as one big function which takes input values ($x$), weights ($w$), and bias ($b$), as inputs, and outputs ($y$).

This big function consists of <b>3 chained functions</b> in total: a multiplication of input values and weights, a sum of these values and bias, as well as a $max$ function as the ReLU activation.


## 9.1. Derivative of activation function

Backpropagate our gradients by calculating derivatives and partial derivatives w.r.t each of our parameters and inputs.

The big function in the context of our neural network, can be loosely interpreted as:

$$
\text{ReLU} \Big( \sum[ \ \text{input} \cdot \text{weights} \ ] + \text{bias} \Big)
$$

Or in the form that matches code more precisely as:

$$
\text{ReLU} \Big( x_0 w_0 + x_1 w_1 + x_2 w_2 + b  \Big)
$$

Rewrite the function to the form that will allow us to determine how to calculate the derivatives more easily:

$$
y = \text{ReLU} \Big( sum \big( mul(x_0 , w_0 ), mul(x_1 , w_1 ), mul(x_2 , w_2 ), b \big) \Big)
$$

Calculate the partial derivative with respect to $w_0$.

$$
\frac{∂}{∂x_0} \bigg[ \text{ReLU} \Big( sum \big( mul(x_0 , w_0 ), mul(x_1 , w_1 ), mul(x_2 , w_2 ), b \big) \Big)  \bigg]
$$

The derivative with respect to the layer’s inputs is not used to update any parameters. It is used to chain to another layer.

We can repeat this to calculate all of the other remaining impacts.

We want to know the impact of a given weight or bias on the loss. Thus, we have to calculate the derivative of the loss function and apply the chain rule with the derivatives of all activation functions and neurons in all of the consecutive layers.

Assume that our neuron receives a gradient of $1$ from the next layer (for demonstration purposes), a value of $1$ won't change the values, that means we can more easily show all of the processes. The color red is used for derivatives.

<center><img src='./image/9-5.png' style='width: 70%'/><font color='gray'><i>Initial gradient (received during backpropagation).</i></font></center>

Recall the derivative of $ReLU()$ w.r.t its input:

$$
f(x) = max(x,0) \quad \to \quad \frac{d}{dx} f(x) = 1 (x>0)
$$

The input value to the $ReLU$ function is $6$, so the derivative equals $1$.

We have to use the chain rule and multiply this derivative with the derivative received from the next layer (which is 1 for the purpose of this example).

In [5]:
# Forward pass
x = [1.0, -2.0, 3.0] # input values
w = [-3.0, -1.0, 2.0] # weights
b = 1.0 # bias

# Multiplying inputs by weights
xw0 = x[0] * w[0]
xw1 = x[1] * w[1]
xw2 = x[2] * w[2]

# Adding weighted inputs and a bias
z = xw0 + xw1 + xw2 + b

# ReLU activation function
y = max(z, 0)

# Backward pass
# The derivative from the next layer
dvalue = 1.0

# Derivative of ReLU and the chain rule
drelu_dz = dvalue * (1. if z > 0 else 0.)
print(drelu_dz)

1.0


<center><img src='./image/9-6.png' style='width: 70%'/><font color='gray'><i>Derivative of the $ReLU$ function and chain rule.</i></font></center>

This results with the derivative of 1 :

<center><img src='./image/9-7.png' style='width: 70%'/><font color='gray'><i>ReLU and chain rule gradient.</i></font></center>



## 9.2. Derivative of a sum of the weighted inputs and bias

Moving backward through our neural network, a sum of the weighted inputs and bias comes immediately before we perform the activation function.

Thus, we must calculate the partial derivative of the sum function, then, using the chain rule, multiply this by the partial derivative of the subsequent, outer, function, which is $ReLU$.

Recall the partial derivative of the sum operation is always 1:

$$
\begin{align}
f(x,y) = x+y \quad \to \quad & \frac{∂}{∂x} f(x,y) = 1\\
& \frac{∂}{∂y} f(x,y) = 1 \
\end{align}
$$

In [6]:
# Forward pass
x = [1.0, -2.0, 3.0] # input values
w = [-3.0, -1.0, 2.0] # weights
b = 1.0 # bias

# Multiplying inputs by weights
xw0 = x[0] * w[0]
xw1 = x[1] * w[1]
xw2 = x[2] * w[2]

# Adding weighted inputs and a bias
z = xw0 + xw1 + xw2 + b

# ReLU activation function
y = max(z, 0)

# Backward pass
# The derivative from the next layer
dvalue = 1.0

# Derivative of ReLU and the chain rule
drelu_dz = dvalue * (1. if z > 0 else 0.)
print(drelu_dz)

# Partial derivatives of the multiplication, the chain rule
dsum_dxw0 = 1
dsum_dxw1 = 1
dsum_dxw2 = 1
dsum_db = 1
drelu_dxw0 = drelu_dz * dsum_dxw0
drelu_dxw1 = drelu_dz * dsum_dxw1
drelu_dxw2 = drelu_dz * dsum_dxw2
drelu_db = drelu_dz * dsum_db

print(drelu_dxw0, drelu_dxw1, drelu_dxw2, drelu_db)

1.0
1.0 1.0 1.0 1.0


<center><img src='./image/9-8.png' style='width: 70%'/><font color='gray'><i>Partial derivative of the sum function w.r.t. the first weighted input.</i></font></center>

This results with a partial derivative of 1 again:

<center><img src='./image/9-9.png' style='width: 70%'/><font color='gray'><i>The sum and chain rule gradient for the first weighted input.</i></font></center>

Perform the same operation with the next weighted input:

<center><img src='./image/9-10.png' style='width: 70%'/><font color='gray'><i>Partial derivative of the sum function w.r.t. the second weighted input.</i></font></center>

Which results with the next calculated partial derivative:

<center><img src='./image/9-11.png' style='width: 70%'/><font color='gray'><i>The sum and chain rule gradient (for the second weighted input).</i></font></center>

And the last weighted input:

<center><img src='./image/9-12.png' style='width: 70%'/><font color='gray'><i>Partial derivative of the sum function w.r.t. the third weighted input.</i></font></center>

<center><img src='./image/9-13.png' style='width: 70%'/><font color='gray'><i>The sum and chain rule gradient (for the third weighted input).</i></font></center>

Then the bias:

<center><img src='./image/9-14.png' style='width: 70%'/><font color='gray'><i>Partial derivative of the sum function w.r.t. the bias.</i></font></center>

<center><img src='./image/9-15.png' style='width: 70%'/><font color='gray'><i>The sum and chain rule gradient (for the bias).</i></font></center>


## 9.3. Derivative of the multiplication of weights and inputs

The derivative for a product is whatever the input is being multiplied by.

$$
\begin{align}
f(x,y) = x \cdot y \quad \to \quad & \frac{∂}{∂x} f(x,y) = y \\
& \frac{∂}{∂y} f(x,y) = x \
\end{align}
$$

In [7]:
# Partial derivatives of the multiplication, the chain rule
dmul_dx0 = w[0]
dmul_dx1 = w[1]
dmul_dx2 = w[2]
dmul_dw0 = x[0]
dmul_dw1 = x[1]
dmul_dw2 = x[2]
drelu_dx0 = drelu_dxw0 * dmul_dx0
drelu_dw0 = drelu_dxw0 * dmul_dw0
drelu_dx1 = drelu_dxw1 * dmul_dx1
drelu_dw1 = drelu_dxw1 * dmul_dw1
drelu_dx2 = drelu_dxw2 * dmul_dx2
drelu_dw2 = drelu_dxw2 * dmul_dw2

print(drelu_dx0, drelu_dw0, drelu_dx1, drelu_dw1, drelu_dx2, drelu_dw2)

-3.0 1.0 -1.0 -2.0 2.0 3.0


<center><img src='./image/9-16.png' style='width: 70%'/><font color='gray'><i>Partial derivative of the multiplication function w.r.t. the first input.</i></font></center>

<center><img src='./image/9-17.png' style='width: 70%'/><font color='gray'><i>The multiplication and chain rule gradient (for the first input).</i></font></center>

The complete set of the activated neuron’s partial derivatives with respect to the inputs, weights and a bias.

<center><img src='./image/9-18.png' style='width: 70%'/><font color='gray'><i>Complete backpropagation graph.</i></font></center>

Recall the equation from the beginning:

$$
\frac{∂}{∂x_0} \bigg[ \text{ReLU} \Big( sum \big( mul(x_0 , w_0 ), mul(x_1 , w_1 ), mul(x_2 , w_2 ), b \big) \Big)  \bigg]
= \frac{d \text{ReLU} }{d sum()} \cdot \frac{∂ sum() }{∂ mul(x_0 , w_0 )} \cdot \frac{∂ mul(x_0 , w_0 )}{∂x_0}
$$

The partial derivative of a neuron’s function, with respect to the weight, is the input related to this weight, and, with respect to the input, is the related weight. The partial derivative of the neuron’s function with respect to the bias is always 1.


## 9.4. Update weight and bias to decrease output

All partial derivatives above combined into a vector, make up our gradients.

In [8]:
dx = [drelu_dx0, drelu_dx1, drelu_dx2] # gradients on inputs
dw = [drelu_dw0, drelu_dw1, drelu_dw2] # gradients on weights
db = drelu_db # gradient on bias...just 1 bias here.

For this single neuron example, we won't need our $dx$.

We will apply these gradients to the weights to hopefully minimize the output.

We apply a negative fraction to this gradient to decrease the final output value, as the gradient shows the direction of the steepest ascent.

In [9]:
print(w, b)

[-3.0, -1.0, 2.0] 1.0


Apply a fraction of the gradients to these values:

In [10]:
w[0] += -0.001 * dw[0]
w[1] += -0.001 * dw[1]
w[2] += -0.001 * dw[2]
b += -0.001 * db
print(w, b)

[-3.001, -0.998, 1.997] 0.999


We slightly changed the weights and bias in such a way to decrease the output intelligently.

Do another forward pass to see the effects:

In [11]:
# Multiplying inputs by weights
xw0 = x[0] * w[0]
xw1 = x[1] * w[1]
xw2 = x[2] * w[2]

# Adding
z = xw0 + xw1 + xw2 + b

# ReLU activation function
y = max(z, 0)

print(y)

5.985


We've successfully decreased this neuron's output from $6.000$ to $5.985$.

It does not make sense to decrease the neuron's output in a real neural network. We do this as a simpler exercise.

We want to decrease the loss value, which is the last calculation in the chain of calculations during the forward pass, and it's the first one to calculate the gradient during the backpropagation.

## 9.5. List of samples and a layer of neurons

### A singular sample

A single neuron of the current layer connects to all of them — they all receive the output of this neuron.

What happen during backpropagation?

Each neuron from the next layer will return a partial derivative of its function w.r.t all of its inputs.

The neuron in the current layer will receive a vector consisting of these derivatives.

To continue backpropagation, we need to sum this vector to be a singular value for a singular neuron.

For a layer of neurons, it'll be a list of these vectors, or a 2D array.

To apply the chain rule, we need to multiply them by the gradient from the subsequent function.

We should perform a sum along the inputs - the first input to all of the neurons, the second input, and so on. So we have to sum columns.

The array of partial derivatives w.r.t all of the inputs equals the array of weights.

Since the array is transposed, we need to sum its rows instead of columns.

Then we calculate the gradient for the next layer in backpropagation.

In [1]:
import numpy as np

# Passed in gradient from the next layer
# for the purpose of this example we're going to use
# a vector of 1s
dvalues = np.array([[1., 1., 1.]])

# We have 3 sets of weights - one set for each neuron
# we have 4 inputs, thus 4 weights
# recall that we keep weights transposed
weights = np.array([[0.2, 0.8, -0.5, 1],
                    [0.5, -0.91 , 0.26, -0.5],
                    [-0.26, -0.27 , 0.17, 0.87]]).T

# sum weights of given input
# and multiply by the passed in gradient for this neuron
dx0 = sum(weights[0] * dvalues[0])
dx1 = sum(weights[1] * dvalues[0])
dx2 = sum(weights[2] * dvalues[0])
dx3 = sum(weights[3] * dvalues[0])

# the gradient of the neuron function with respect to inputsm
dinputs = np.array([dx0, dx1, dx2, dx3])

print(dinputs)

[ 0.44 -0.38 -0.07  1.37]


We can achieve the same result by using the `np.dot` function. 

Recall, we have one partial derivative for each neuron and multiply it by the neuron’s partial derivative with respect to its input.

Then, multiply each of these gradients with each of the partial derivatives that are related to this neuron’s inputs.


In [13]:
import numpy as np

# Passed in gradient from the next layer
# for the purpose of this example we're going to use
# a vector of 1s
dvalues = np.array([[1., 1., 1.]])

# We have 3 sets of weights - one set for each neuron
# we have 4 inputs, thus 4 weights
# recall that we keep weights transposed
weights = np.array([[0.2, 0.8, -0.5, 1],
                    [0.5, -0.91, 0.26, -0.5 ],
                    [-0.26, -0.27, 0.17, 0.87 ]]).T

# sum weights of given input
# and multiply by the passed in gradient for this neuron
dinputs = np.dot(dvalues[ 0 ], weights.T)

print (dinputs)

[ 0.44 -0.38 -0.07  1.37]


### A batch of samples

In [14]:
import numpy as np

# Passed in gradient from the next layer
# for the purpose of this example we're going to use
# an array of an incremental gradient values
dvalues = np.array([[1., 1., 1.],
                    [2., 2., 2.],
                    [3., 3., 3.]])

# We have 3 sets of weights - one set for each neuron
# we have 4 inputs, thus 4 weights
# recall that we keep weights transposed
weights = np.array([[0.2, 0.8, -0.5, 1],
                    [0.5, -0.91, 0.26, -0.5],
                    [-0.26, -0.27, 0.17, 0]]).T

# sum weights of given input
# and multiply by the passed in gradient for this neuron
dinputs = np.dot(dvalues, weights.T)

print (dinputs)

[[ 0.44 -0.38 -0.07  0.5 ]
 [ 0.88 -0.76 -0.14  1.  ]
 [ 1.32 -1.14 -0.21  1.5 ]]


In [15]:
import numpy as np

# Passed in gradient from the next layer for the purpose of this example 
# We're going to use an array of an incremental gradient values
dvalues = np.array([[1., 1., 1.],
                    [2., 2., 2.],
                    [3., 3., 3.]])

# We have 3 sets of inputs - samples
inputs = np.array([[1, 2, 3, 2.5],
                [2., 5., -1., 2],
                [-1.5, 2.7, 3.3, -0.8]])

# Sum weights of given input and multiply by the passed in gradient for this neuron
dweights = np.dot(inputs.T, dvalues)

print(dweights)

[[ 0.5  0.5  0.5]
 [20.1 20.1 20.1]
 [10.9 10.9 10.9]
 [ 4.1  4.1  4.1]]


For the biases and derivatives with respect to them, the derivatives come from the sum operation and always equal 1, multiplied by the incoming gradients to apply the chain rule.

In [16]:
import numpy as np

# Passed in gradient from the next layer for the purpose of this example
# we're going to use an array of an incremental gradient values
dvalues = np.array([[1., 1., 1.],
                    [2., 2., 2.],
                    [3., 3., 3.]])

# One bias for each neuron biases are the row vector with a shape (1, neurons)
biases = np.array([[2, 3, 0.5]])

# dbiases - sum values, do this over samples (first axis), keepdims
# since this by default will produce a plain list - we explained this in the chapter 4
dbiases = np.sum(dvalues, axis = 0 , keepdims = True )

print(dbiases)

[[6. 6. 6.]]


### Derivative of the ReLU function

It equals 1 if the input is greater than 0 and 0 otherwise.

The layer passes its outputs through the $ReLU()$ activation during the forward pass.

For the backward pass, $ReLU()$ receives a gradient of the same shape.

The derivative of the ReLU function will form an array of the same shape, filled with 1 when the related input is greater than 0, and 0 otherwise.

To apply the chain rule, we need to multiply this array with the gradients of the following function.

In [17]:
import numpy as np

# Example layer output (4 neurons)
z = np.array([[1, 2, -3, -4],
              [2, -7, -1, 3],
              [-1, 2, 5, -1]])

dvalues = np.array([[1, 2, 3, 4],
                    [5, 6, 7, 8],
                    [9, 10, 11, 12]])

# ReLU activation's derivative
drelu = np.zeros_like(z)
drelu[z > 0] = 1
print(drelu)

# The chain rule
drelu *= dvalues

print(drelu)

[[1 1 0 0]
 [1 0 0 1]
 [0 1 1 0]]
[[ 1  2  0  0]
 [ 5  0  0  8]
 [ 0 10 11  0]]


We can simplify this operation.

In [18]:
import numpy as np

# Example layer output
z = np.array([[1, 2, -3, -4],
              [2, -7, -1, 3],
              [-1, 2, 5, -1]])

dvalues = np.array([[1, 2, 3, 4],
                    [5, 6, 7, 8],
                    [9, 10, 11, 12]])

# ReLU activation's derivative with the chain rule applied
drelu = dvalues.copy()    # ensures that we don't modify it during the ReLU derivative calculation.
drelu[z <= 0] = 0

print(drelu)

[[ 1  2  0  0]
 [ 5  0  0  8]
 [ 0 10 11  0]]


Let’s combine the forward and backward pass of a single neuron with a full layer and batch-based partial derivatives. We’ll minimize ReLU’s output, once again, only for this example:

In [2]:
import numpy as np

# Passed in gradient from the next layer for the purpose of this example
# We're going to use an array of an incremental gradient values
dvalues = np.array([[1., 1., 1.],
                    [2., 2., 2.],
                    [3., 3., 3.]])

# We have 3 sets of inputs - samples
inputs = np.array([[1, 2, 3, 2.5],
                   [2., 5., -1., 2],
                   [-1.5, 2.7, 3.3, -0.8]])

# We have 3 sets of weights - one set for each neuron
# we have 4 inputs, thus 4 weights
# recall that we keep weights transposed
weights = np.array([[0.2, 0.8, -0.5, 1],
                    [0.5, -0.91, 0.26, -0.5],
                    [-0.26, -0.27, 0.17, 0.87]]).T

# One bias for each neuron
# biases are the row vector with a shape (1, neurons)
biases = np.array([[2, 3, 0.5]])

# Forward pass
layer_outputs = np.dot(inputs, weights) + biases # Dense layer
relu_outputs = np.maximum(0, layer_outputs) # ReLU activation

# Let's optimize and test backpropagation here
# ReLU activation - simulates derivative with respect to input values
# from next layer passed to current layer during backpropagation
drelu = dvalues.copy()
drelu[layer_outputs <= 0] = 0

# Dense layer
# dinputs - multiply by weights
dinputs = np.dot(drelu, weights.T)

# dweights - multiply by inputs
dweights = np.dot(inputs.T, drelu)

# dbiases - sum values, do this over samples (first axis), keepdims
# since this by default will produce a plain list - we explained this in the chapter 4
dbiases = np.sum(drelu, axis = 0 , keepdims = True)

# Update parameters
weights += - 0.001 * dweights
biases += - 0.001 * dbiases

print(weights)
print(biases)

[[ 0.1995  0.5035 -0.2605]
 [ 0.7799 -0.9201 -0.2901]
 [-0.5109  0.2471  0.1591]
 [ 0.9959 -0.5001  0.8659]]
[[1.994 2.996 0.494]]


Update the dense layer and ReLU activation code with a `backward` method (for backpropagation).

In [20]:
# Dense layer
class Layer_Dense :
    # Layer initialization
    def __init__ (self, inputs, neurons):
        self.weights = 0.01 * np.random.randn(inputs, neurons)
        self.biases = np.zeros((1, neurons))

    # Forward pass
    def forward(self, inputs):
        self.output = np.dot(inputs, self.weights) + self.biases
        # To remember what the inputs were (needed when calculating the partial derivative w.r.t weights during backpropagation)
        self.inputs = inputs
        
    # Backward pass
    def backward(self, dvalues):
        # Gradients on parameters
        self.dweights = np.dot(self.inputs.T, dvalues)
        self.dbiases = np.sum(dvalues, axis = 0, keepdims = True)

        # Gradient on values
        self.dinputs = np.dot(dvalues, self.weights.T)


# ReLU activation
class Activation_ReLU :
    # Forward pass
    def forward(self, inputs):
        # Remember input values
        self.inputs = inputs
        self.output = np.maximum(0, inputs)

    # Backward pass
    def backward(self, dvalues):
        # Since we need to modify the original variable, let's make a copy of the values first
        self.dinputs = dvalues.copy()

        # Zero gradient where input values were negative
        self.dinputs[self.inputs <= 0 ] = 0

## 9.6. Categorical Cross-Entropy loss derivative

As stated in chapter 5, the Categorical Cross-Entropy loss function's formula is:

$$
L_i = \ \text{log} \ \big(  \hat{y}_{i,k} \big) \qquad \text{where }  k \text{ is an index of "true" probability}
$$

where $L_i$ denotes sample loss value, $i^{th}$ sample in a set, $k$ - index of the target label (ground-true label), $y$ - target values and $\hat y$ predicted values.

All we need is the output of the Softmax activation function at the index of the correct class.

To calculate partial derivatives with respect to each of the inputs, we need an equation that takes all of them as parameters, thus the choice to use the full equation.

For the purpose of the derivative calculation, we use the full equation mentioned back in chapter 5:

$$
L_i = - \sum_j y_{i,j} \ \text{log} \ \big(  \hat{y}_{i,j} \big)
$$

where $L_i$ denotes sample loss value, $i^{th}$ sample in a set, $j$ - label/output index, $y$ - target values and $\hat y$ predicted values.

Recall,

$$
f(x) = \text{log} \big( h(x) \big) \quad \to \quad f'(x) = \frac{1}{h(x)} \cdot h'(x)\\
f(x) = \text{log} (x) \quad \to \quad \frac{d}{dx} f(x) = \frac{d}{dx} \text{log} (x) =   \frac{1}{x} \cdot \frac{d}{dx} x = \frac{1}{x} \cdot 1 = \frac{1}{x}
$$

First, let’s define the gradient equation as the partial derivative of the loss function w.r.t each of its inputs.

$$
\frac{∂ L_i}{∂ \hat{y}_{i,j}} = \frac{∂ }{∂ \hat{y}_{i,j}} \Big[  - \sum_j y_{i,j} \ \text{log} \ \big(  \hat{y}_{i,j} \big)  \Big] =   - \sum_j y_{i,j} \cdot \frac{∂ }{∂ \hat{y}_{i,j}}  \text{log} \big(  \hat{y}_{i,j} \big)  =   - \sum_j y_{i,j} \cdot \frac{1}{\hat{y}_{i,j}} \cdot \frac{∂ }{∂ \hat{y}_{i,j}}  \hat{y}_{i,j}  =   - \sum_j y_{i,j} \cdot \frac{1}{\hat{y}_{i,j}} \cdot 1  =  - \sum_j \frac{y_{i,j}}{\hat{y}_{i,j}}  =  - \frac{y_{i,j}}{\hat{y}_{i,j}}
$$

The derivative of this loss function with respect to its inputs (predicted values at the $i^{th}$ sample, since we are interested in a gradient with respect to the predicted values) equals the negative ground-truth vector, divided by the vector of the predicted values (which is also the output vector of the softmax function).


## 9.7. Categorical Cross-Entropy loss derivative code implementation

Add a backward method to the `Loss_CategoricalCrossentropy` class, and pass the array of predictions and the array of true values into it and calculate the negated division of them:

In [21]:
# Common loss class
class Loss :
    
    # Calculates the data and regularization losses given model output and ground truth values
    def calculate ( self , output , y ):
        
        # Calculate sample losses
        sample_losses = self.forward(output, y)
        
        # Calculate mean loss
        data_loss = np.mean(sample_losses)
        
        # Return loss
        
        return data_loss

    
# Cross-entropy loss
class Loss_CategoricalCrossentropy(Loss):
    
    # Forward pass
    def forward(self, y_pred, y_true):
        
        # Number of samples in a batch
        samples = len(y_pred)
        
        # Clip data to prevent division by 0
        # Clip both sides to not drag mean towards any value
        y_pred_clipped = np.clip(y_pred, 1e-7, 1 - 1e-7)
        
        # Probabilities for target values - only if categorical labels
        if len(y_true.shape) == 1:
            correct_confidences = y_pred_clipped[ range(samples), y_true]
        
        # Mask values - only for one-hot encoded labels
        elif len(y_true.shape) == 2:
            correct_confidences = np.sum( y_pred_clipped*y_true, axis=1)
        
        # Losses
        negative_log_likelihoods = -np.log(correct_confidences)
        
        return negative_log_likelihoods

    # Backward pass
    def backward (self, dvalues, y_true):

        # Number of samples
        samples = len(dvalues)

        # Number of labels in every sample
        # We'll use the first sample to count them
        labels = len(dvalues[0])

        # If labels are sparse, turn them into one-hot vector
        if len(y_true.shape) == 1 :
            y_true = np.eye(labels)[y_true]

        # Calculate gradient
        self.dinputs = -y_true/dvalues

        # Normalize gradient
        self.dinputs = self.dinputs/samples


Optimizers sum all of the gradients related to each weight and bias before multiplying them by the learning rate (or some other factor). This means that the more samples we have in a dataset, the more gradient sets we’ll receive at this step, and the bigger this sum will become. As a consequence, we’ll have to adjust the learning rate according to each set of samples.

To solve this problem, we can divide all of the gradients by the number of samples. A sum of elements divided by a count of them is their mean value (the optimizer will perform the sum). Thus, we will effectively normalize the gradients and make their sum’s magnitude invariant to the number of samples.


## 9.8. Softmax activation derivative

The partial derivative of the Softmax function is a bit more complicated task than the derivative of the Categorical Cross-Entropy loss.

Recall, the equation of the Softmax activation function and define the derivative:

$$
S_{i,j} = \frac{e^{z_{i,j}}}{\sum_{l=1}^{L} e^{z_{i,l}}} \quad \to \quad \frac{∂ S_{i,j}}{∂z_{i,k}} = \frac{∂ \frac{e^{z_{i,j}}}{\sum_{l=1}^{L} e^{z_{i,l}}}}{∂z_{i,k}}
$$

where $S_{i,j}$ denotes $j^{th}$ Softmax’s output of $i^{th}$ sample, $z$ - input array which is a list of input vectors (output vectors from the previous layer), $z_{i,j}$ - $j^{th}$ Softmax’s input of $i^{th}$ sample, $L$ - number of inputs, $z_{i,k}$ - $k^{th}$ Softmax’s input of $i^{th}$ sample.

As described in chapter 4, the Softmax function equals the exponential input divided by the sum of all exponentiated inputs.

Each input to the Softmax impacts each of the outputs, and we need to calculate the partial derivative of each output with respect to each input.

If we calculate the impact of one list on the other list, we’ll receive a matrix of values as a result.

It is the Jacobian matrix of the vectors.

Recall, the derivative of the division operation:

$$
f(x) = \frac{g(x)}{h(x)} \quad \to \quad f'(x) = \frac{ g'(x) \cdot h(x) - g(x) \cdot h'(x)}{\big[ h(x) \big]^2}
$$

Thus,

$$
\frac{∂ S_{i,j}}{∂z_{i.k}} = \frac{∂ \frac{e^{z_{i,j}}}{\sum_{l=1}^{L} e^{z_{i,l}}}}{∂z_{i.k}} 
= \frac{ \frac{∂}{∂z_{i,k}} e^{z_i,j} \cdot \sum_{l=1}^{L} e^{z_{i,l}} \ - \   e^{z_i,j} \cdot \frac{∂}{∂z_{i,k}} \sum_{l=1}^{L} e^{z_{i,l}} }{ \big[ \sum_{l=1}^{L} e^{z_{i,l}} \big]^2}
$$

Recall,

$$
\frac{d}{dn} e^n = e^n \cdot \frac{d}{dn} n = e^n \cdot 1 = e^n
$$

The RHS of the numerator:

$$
\frac{∂}{∂z_{i,k}} \sum_{l=1}^{L} e^{z_{i,l}}
$$

The range $1 \cdot \cdot \cdot L$ contains $k$ ($k$ is one of the indices from this range) exactly once and then, in this case, the derivative is going to equal $e$ to the power of the $z_{i,k}$ (as $j$ equals $k$) and $0$ otherwise (when $j$ does not equal $k$ as $z_{i,l}$ won’t contain $z_{i,k}$ and will be treated as a constant — The derivative of the constant equals $0$):

$$
\begin{align}
\frac{∂}{∂z_{i,k}} \sum_{l=1}^{L} e^{z_{i,l}} & = \frac{∂}{∂z_{i,k}} e^{z_{i, 1}} + \frac{∂}{∂z_{i,k}} e^{z_{i, 2}}  + \cdots + \frac{∂}{∂z_{i,k}} e^{z_{i, k}} + \cdots + \frac{∂}{∂z_{i,k}} e^{z_{i, L-1}} + \frac{∂}{∂z_{i,k}} e^{z_{i, L}}\\
& = 0 + 0 + \cdots + e^{z_{i ,k} } + \cdots + 0 + 0 = e^{z_{i ,k} } \
\end{align}
$$

The LHS of the numerator:

$$
\frac{∂}{∂z_{i,k}}  e^{z_{i,j}}
$$

It can become either $0$ if $j \neq k$ or $e$ to the power of the $z_{i, k}$ if $j=k$.

<b>Starting from this step, we need to calculate the derivatives separately for both cases.</b>

In the case of $i=j$:

$$
\begin{align}
\frac{∂ S_{i, \ j}}{∂z_{i, \ k}} & = \frac{ e^{z_{i, \ j}} \cdot \sum_{l=1}^{L} e^{z_{i, \ l}} - e^{z_{i, \ j}} \cdot e^{z_{i, \ k}}}{\big[ \sum_{l=1}^{L} e^{z_{i, \ l}} \big]^2} = \frac{e^{z_{i, \ j}} \cdot \big(\sum_{l=1}^{L} e^{z_{i, \ l}} - e^{z_{i, \ k}}\big)}{\sum_{l=1}^{L} e^{z_{i, \ l}} \cdot  \sum_{l=1}^{L} e^{z_{i, \ l}} } \\
& = \frac{e^{z_{i, \ j}} }{\sum_{l=1}^{L} e^{z_{i, \ l}}}  \cdot \frac{\sum_{l=1}^{L} e^{z_{i, \ l}} - e^{z_{i, \ k}}}{\sum_{l=1}^{L} e^{z_{i, \ l}}} = \frac{e^{z_{i, \ j}}}{\sum_{l=1}^{L} e^{z_{i, \ l}}} \cdot \bigg(\frac{\sum_{l=1}^{L} e^{z_{i, \ l}}}{\sum_{l=1}^{L} e^{z_{i, \ l}}} - \frac{e^{z_{i, \ k}}}{\sum_{l=1}^{L} e^{z_{i, \ l}}} \bigg) = S_{i, \ j} \cdot ( 1 - S_{i, \ k}) \
\end{align}
$$

In the case of $i \neq j$:

$$
\begin{align}
\frac{∂ S_{i,j}}{∂z_{i,k}} & = \frac{ 0 \cdot \sum_{l=1}^{L} e^{z_{i, \ l}} \ - \ e^{z_{i, \ j}} \cdot e^{z_{i, \ k}}  }{ \big[ \sum_{l=1}^{L} e^{z_{i, \ l}} \big]^2} = \frac{  - \ e^{z_{i, \ j}} \cdot e^{z_{i, \ k}}  }{ \big[ \sum_{l=1}^{L} e^{z_{i, \ l}} \big]^2} = \frac{  - \ e^{z_{i, \ j}} \cdot e^{z_{i, \ k}}  }{  \sum_{l=1}^{L} e^{z_{i, \ l}} \cdot \sum_{l=1}^{L} e^{z_{i, \ l}} } = - \frac{  \ e^{z_{i, \ j}}  }{  \sum_{l=1}^{L} e^{z_{i, \ l}} }  \cdot \frac{ e^{z_{i, \ k}}  }{ \sum_{l=1}^{L} e^{z_{i, \ l}} } = - S_{i, \ j } \cdot S_{i, \ k }\
\end{align}
$$

As a summary, the solution of the derivative of the Softmax function with respect to its inputs is:

$$
\frac{∂ S_{i,j}}{∂z_{i, \ k}}  =
\begin{cases}
S_{i, \ j} \cdot ( 1 - S_{i, \ k})  , & j = k \\
- S_{i, \ j } \cdot S_{i, \ k }, & j ≠ k
\end{cases}
$$

Now, we have 2 separate equations to code and use in different cases, which isn't very convenient for the speed of calculations.

We rewrite:

$$
\frac{∂ S_{i, \ j}}{∂z_{i, \ k}}  =
\begin{cases}
S_{i, \ j} \cdot ( 1 - S_{i, \ k })  , & j = k \\
S_{i, \ j} \cdot ( 0 - S_{i, \ k }) ,  & j ≠ k
\end{cases}
$$

There exists Kronecker delta function whose equation is:

$$
\delta _{i, \ j}  =
\begin{cases}
1 , & i = j \\
0 , & i ≠ j
\end{cases}
$$

Apply it to simplify the equation further:

$$
\frac{∂ S_{i, \ j}}{∂z_{i, \ k}}  = S_{i, \ j} \cdot ( \delta _{j, \ k} - S_{i, \ k })
$$

To make it a little bit easier to implement in Python using NumPy, we have the final math solution to the derivative of the Softmax function's outputs with respect to each of its inputs.

$$
\frac{∂ S_{i, \ j}}{∂z_{i, \ k}}  = S_{i, \ j} \cdot ( \delta _{j, \ k} - S_{i, \ k })
= S_{i, \ j} \delta _{j, \ k} - S_{i, \ j}  S_{i, \ k }
$$


## 9.9. Softmax activation derivative code implementation

We can code the solution using just two NumPy functions.

In our case, the Jacobian matrix is an array of partial derivatives in all of the combinations of both input vectors.

In section 9.8, we calculate the partial derivatives of every output of the Softmax function w.r.t each input separately, as each input influences each output due to the normalization process, which takes the sum of all the exponentiated inputs.

In [22]:
import numpy as np

softmax_output = [0.7, 0.1, 0.2]
softmax_output = np.array(softmax_output).reshape(-1, 1)

# First terms
S_ijd_jk = softmax_output * np.eye(softmax_output.shape[0])
S_ijd_jk = np.diagflat(softmax_output)

# Second terms
S_ijS_ik = np.dot(softmax_output, softmax_output.T)

# The Jacobian matrix
d_Sij_d_zik = S_ijd_jk - S_ijS_ik
print(d_Sij_d_zik)


[[ 0.21 -0.07 -0.14]
 [-0.07  0.09 -0.02]
 [-0.14 -0.02  0.16]]


On a batch of samples, this operation returns a list if the Jacobian matrices, which forms a 3D matrix. We can visualize it as a column whose levels are Jacobian matrices being the sample-wise gradient of the Softmax function.

This raises a question — if sample-wise gradients are the Jacobian matrices, how do we perform the chain rule with the gradient back-propagated from the loss function, since it’s a vector for each sample? Also, what do we do with the fact that the previous layer, which is the Dense layer, will expect the gradients to be a 2D array? Currently, we have a 3D array of the partial derivatives — a list of the Jacobian matrices. The derivative of the Softmax function with respect to any of its inputs returns a vector of partial derivatives (a row from the Jacobian matrix), as this input influences all the outputs, thus also influencing the partial derivative for each of them. We need to sum the values from these vectors so that each of the inputs for each of the samples will return a single partial derivative value instead. Because each input influences all of the outputs, the returned vector of the partial derivatives has to be summed up for the final partial derivative with respect to this input. We can perform this operation on each of the Jacobian matrices directly, applying the chain rule at the same time (applying the gradient from the loss function) using `np.dot()` — For each sample, it’ll take the row from the Jacobian matrix and multiply it by the corresponding value from the loss function’s gradient. As a result, the dot product of each of these vectors and values will return a singular value, forming a vector of the partial derivatives sample-wise and a 2D array (a list of the resulting vectors) batch-wise.

Let’s code the solution:

In [23]:
# Softmax activation
class Activation_Softmax:
    
    # Forward pass
    def forward(self, inputs):
        
        # Get unnormalized probabilities
        exp_values = np.exp(inputs - np.max(inputs, axis=1, keepdims=True))
        
        # Normalize them for each sample
        probabilities = exp_values / np.sum(exp_values, axis=1, keepdims=True)
        self.output = probabilities
        
        
    # Backward pass
    def backward(self, dvalues):

        # Create uninitialized array
        self.dinputs = np.empty_like(dvalues)

        # Enumerate outputs and gradients
        for index, (single_output, single_dvalues) in enumerate (zip(self.output, dvalues)):
            
            # Flatten output array
            single_output = single_output.reshape(-1, 1)

            # Calculate Jacobian matrix of the output
            jacobian_matrix = np.diagflat(single_output) - np.dot(single_output, single_output.T)

            # Calculate sample-wise gradient and add it to the array of sample gradients
            self.dinputs[index] = np.dot(jacobian_matrix, single_dvalues)  

## 9.10. Common Categorical Cross-Entropy loss and Softmax activation derivative

We already calculated the partial derivatives of the Categorial Cross-Entropy loss and Softmax activation functions.

Apply the chain rule to calculate the partial derivative of the Categorical Cross-Entropy loss function w.r.t the Softmax function inputs.

Note that the inputs to the loss function $\hat{y}_{ i, \ j}$ are the outputs of the activation function $S_{i, \ j}$:

$$
\hat{y}_{ i, \ j} = S_{i, \ j}
$$

Let's define this derivative by applying the chain rule:

$$
\frac{∂ L_i }{∂z_{i, \ k}}  = \frac{∂ L_i }{∂ \hat{y}_{i, \ j}}  \cdot \frac{∂ S_{i, \ j}}{∂z_{i, \ k}} = \frac{∂ L_i }{∂ \hat{y}_{i, \ j}}  \cdot \frac{∂ \hat{y}_{i, \ j}}{∂ z_{i, \ k}} =  - \sum_j \frac{y_{i, \ j}}{\hat{y}_{i, \ j}} \cdot \frac{ ∂ \hat{y}_{i, \ j}  }{ ∂ z_{i, \ k} }
$$

Recall from section 9.8, the derivative of the Softmax function with respect to its inputs is:

$$
\frac{∂ S_{i,j}}{∂z_{i, \ k}}  =
\begin{cases}
S_{i, \ j} \cdot ( 1 - S_{i, \ k})  , & j = k \\
- S_{i, \ j } \cdot S_{i, \ k }, & j ≠ k
\end{cases}
$$

Let's substitute $S_{i, \ j }$ with $\hat{y}_{i, \ j}$ we have:

$$
\frac{∂ \hat{y}_{i,j}}{∂z_{i, \ k}}  =
\begin{cases}
\hat{y}_{i, \ j} \cdot ( 1 - \hat{y}_{i, \ k})  , & j = k \\
- \hat{y}_{i, \ j } \cdot \hat{y}_{i, \ k }, & j ≠ k
\end{cases}
$$

We have to split the current partial derivative following these cases (when $j=k$ and $j \neq k$):

$$
\begin{align}
\frac{∂ L_i }{∂z_{i, \ k}}  & = - \sum_j \frac{y_{i, \ j}}{\hat{y}_{i, \ j}} \cdot \frac{ ∂ \hat{y}_{i, \ j}  }{ ∂ z_{i, \ k}  } \\
& = - \frac{y_{i, \ k}}{\hat{y}_{i, \ k}} \cdot \frac{ ∂ \hat{y}_{i, \ k}  }{ ∂ z_{i, \ k}  } - \sum_{j ≠ k} \frac{y_{i, \ j}}{\hat{y}_{i, \ j}} \cdot \frac{ ∂ \hat{y}_{i, \ j}  }{ ∂ z_{i, \ k}  } \\
& =  - \frac{y_{i, \ k}}{\hat{y}_{i, \ k}} \cdot \hat{y}_{i, \ k} \cdot (1- \hat{y}_{i, \ k}) - \sum_{j ≠ k} \frac{y_{i, \ j}}{\hat{y}_{i, \ j}} ( - \hat{y}_{i, \ j} \ \hat{y}_{i, \ k}) \\
& = - y_{i, \ k} \cdot (1- \hat{y}_{i, \ k}) + \sum_{j ≠ k} y_{i, \ j} \ \hat{y}_{i, \ k} \\
& = - y_{i, \ k} + y_{i, \ k} \ \hat{y}_{i, \ k} + \sum_{j ≠ k} y_{i, \ j} \ \hat{y}_{i, \ k} \\
& = - y_{i, \ k} + \sum_{j} y_{i, \ j} \ \hat{y}_{i, \ k} \
\end{align}
$$

Now the sum operator iterates over all of the possible values of $j$ and, since we know that $y_{i, \ j}$ for each $i$ is the one-hot encoded vector of ground-truth values, the sum of all of its elements equas 1.

In other words, this sum will multiply $0$ by the $\hat{y}_{i, \ k}$ except for a single situation, the true label, where it'll muliply $1$ by this value.

Thus, we can further simplify:

$$
\frac{∂ L_i}{∂ \hat{y}_{i,j}} = - y_{i, \ k} + \sum_{j} y_{i, \ j} \ \hat{y}_{i, \ k} = - y_{i, \ k} + \hat{y}_{i, \ k}  = \hat{y}_{i, \ k}  - y_{i, \ k} \
$$

Full solution:

$$
\begin{align}
\frac{∂ L_i}{∂ \hat{y}_{i,j}} &  = \frac{∂ }{∂ \hat{y}_{i,j}} \Big[  - \sum_j y_{i,j} \ \text{log} \ \big(  \hat{y}_{i,j} \big)  \Big] = \frac{∂ L_i }{∂ \hat{y}_{i, \ j}}  \cdot \frac{∂ \hat{y}_{i, \ j}}{∂ z_{i, \ k}} \\
& = - \sum_j \frac{y_{i, \ j}}{\hat{y}_{i, \ j}} \cdot \frac{ ∂ \hat{y}_{i, \ j}  }{ ∂ z_{i, \ k}  } = - \frac{y_{i, \ k}}{\hat{y}_{i, \ k}} \cdot \frac{ ∂ \hat{y}_{i, \ k}  }{ ∂ z_{i, \ k}  } - \sum_{j ≠ k} \frac{y_{i, \ j}}{\hat{y}_{i, \ j}} \cdot \frac{ ∂ \hat{y}_{i, \ j}  }{ ∂ z_{i, \ k}  } \\
& = - \frac{y_{i, \ k}}{\hat{y}_{i, \ k}} \cdot \hat{y}_{i, \ k} \cdot (1- \hat{y}_{i, \ k}) - 
\sum_{j ≠ k} \frac{y_{i, \ j}}{\hat{y}_{i, \ j}} ( - \hat{y}_{i, \ j} \ \hat{y}_{i, \ k}) = - y_{i, \ k} \cdot (1- \hat{y}_{i, \ k}) + \sum_{j ≠ k} y_{i, \ j} \ \hat{y}_{i, \ k} \\
& = - y_{i, \ k} + y_{i, \ k} \ \hat{y}_{i, \ k}) + \sum_{j ≠ k} y_{i, \ j} \ \hat{y}_{i, \ k}  =  - y_{i, \ k} + \sum_{j} y_{i, \ j} \ \hat{y}_{i, \ k}  \\
& = - y_{i, \ k} + \hat{y}_{i, \ k}  = \hat{y}_{i, \ k}  - y_{i, \ k} \
\end{align}
$$

We can see that, when we apply the chain rule to both partial derivatives, the whole equation simplifies significantly to the subtraction of the predicted and ground truth values. It is also multiple times faster to compute.

## 9.11. Common Categorical Cross-Entropy loss and Softmax activation derivative - code implementation

No changes in the forward pass - we still need to perform it on the activation function to receive the outputs and then on the loss function to calculate the loss value.

We'll code the solution as a separate class, which initializes both the Softmax activation and the Categorical Cross-Entropy objects, calling their forward methods respectively during the forward pass.

For backpropagation, we create the backward step to calculates the combined gradient of the loss and activation functions.

In [24]:
# Softmax classifier - combined Softmax activation and cross-entropy loss for faster backward step
class Activation_Softmax_Loss_CategoricalCrossentropy():

    # Creates activation and loss function objects
    def __init__ ( self ):
        self.activation = Activation_Softmax()
        self.loss = Loss_CategoricalCrossentropy()

    # Forward pass
    def forward ( self , inputs , y_true ):
        # Output layer's activation function
        self.activation.forward(inputs)

        # Set the output
        self.output = self.activation.output

        # Calculate and return loss value
        return self.loss.calculate(self.output, y_true)

    # Backward pass
    def backward ( self , dvalues , y_true ):
        # Number of samples
        samples = len(dvalues)

        # If labels are one-hot encoded, turn them into discrete values
        if len (y_true.shape) == 2 :
            y_true = np.argmax(y_true, axis = 1 )

        # Copy so we can safely modify
        self.dinputs = dvalues.copy()
        
        # Calculate gradient
        self.dinputs[range(samples), y_true] -= 1
        
        # Normalize gradient
        self.dinputs = self.dinputs / samples


## 9.12. Summary

In [25]:
# Common loss class
class Loss :
    
    # Calculates the data and regularization losses given model output and ground truth values
    def calculate ( self , output , y ):
        
        # Calculate sample losses
        sample_losses = self.forward(output, y)
        
        # Calculate mean loss
        data_loss = np.mean(sample_losses)
        
        # Return loss
        
        return data_loss


    
# Softmax activation
class Activation_Softmax:

    # Forward pass
    def forward(self, inputs):
        
        # Remember input values
        self.inputs = inputs

        # Get unnormalized probabilities
        exp_values = np.exp(inputs - np.max(inputs, axis=1, keepdims=True))
        
        # Normalize them for each sample
        probabilities = exp_values / np.sum(exp_values, axis=1, keepdims=True)
        self.output = probabilities

    # Backward pass
    def backward(self, dvalues):
        
        # Create uninitialized array
        self.dinputs = np.empty_like(dvalues)

        # Enumerate outputs and gradients
        for index, (single_output, single_dvalues) in enumerate(zip(self.output, dvalues)):
            
            # Flatten output array
            single_output = single_output.reshape(-1, 1)
            
            # Calculate Jacobian matrix of the output
            jacobian_matrix = np.diagflat(single_output) - np.dot(single_output, single_output.T)

            # Calculate sample-wise gradient and add it to the array of sample gradients
            self.dinputs[index] = np.dot(jacobian_matrix, single_dvalues)


            
# Cross-entropy loss
class Loss_CategoricalCrossentropy(Loss):

    # Forward pass
    def forward(self, y_pred, y_true):

        # Number of samples in a batch
        samples = len(y_pred)

        # Clip data to prevent division by 0
        # Clip both sides to not drag mean towards any value
        y_pred_clipped = np.clip(y_pred, 1e-7, 1 - 1e-7)

        # Probabilities for target values - only if categorical labels
        if len(y_true.shape) == 1:
            correct_confidences = y_pred_clipped[range(samples), y_true]

        # Mask values - only for one-hot encoded labels
        elif len(y_true.shape) == 2:
            correct_confidences = np.sum(y_pred_clipped * y_true, axis=1)

        # Losses
        negative_log_likelihoods = -np.log(correct_confidences)
        return negative_log_likelihoods

    # Backward pass
    def backward(self, dvalues, y_true):

        # Number of samples
        samples = len(dvalues)
        
        # Number of labels in every sample
        # We'll use the first sample to count them
        labels = len(dvalues[0])

        # If labels are sparse, turn them into one-hot vector
        if len(y_true.shape) == 1:
            y_true = np.eye(labels)[y_true]

        # Calculate gradient
        self.dinputs = -y_true / dvalues
        
        # Normalize gradient
        self.dinputs = self.dinputs / samples

        
        
# Softmax classifier - combined Softmax activation and cross-entropy loss for faster backward step
class Activation_Softmax_Loss_CategoricalCrossentropy():

    # Creates activation and loss function objects
    def __init__(self):
        self.activation = Activation_Softmax()
        self.loss = Loss_CategoricalCrossentropy()

    # Forward pass
    def forward(self, inputs, y_true):
        # Output layer's activation function
        self.activation.forward(inputs)
        
        # Set the output
        self.output = self.activation.output
        
        # Calculate and return loss value
        return self.loss.calculate(self.output, y_true)

    # Backward pass
    def backward(self, dvalues, y_true):

        # Number of samples
        samples = len(dvalues)

        # If labels are one-hot encoded, turn them into discrete values
        if len(y_true.shape) == 2:
            y_true = np.argmax(y_true, axis=1)

        # Copy so we can safely modify
        self.dinputs = dvalues.copy()
        
        # Calculate gradient
        self.dinputs[range(samples), y_true] -= 1
        
        # Normalize gradient
        self.dinputs = self.dinputs / samples

Let's make up an output of the Softmax function and some target values.

Next, backpropagate them using both solutions (the combined backward, through both of the functions separately) to test that they should return the same values.

In [26]:
import numpy as np
import nnfs

nnfs.init()

softmax_outputs = np.array([[ 0.7 , 0.1 , 0.2 ],
                            [ 0.1 , 0.5 , 0.4 ],
                            [ 0.02 , 0.9 , 0.08 ]])
class_targets = np.array([ 0 , 1 , 1 ])

# Combined method
softmax_loss = Activation_Softmax_Loss_CategoricalCrossentropy()
softmax_loss.backward(softmax_outputs, class_targets)
dvalues1 = softmax_loss.dinputs

# Separate method
activation = Activation_Softmax()
activation.output = softmax_outputs
loss = Loss_CategoricalCrossentropy()
loss.backward(softmax_outputs, class_targets)
activation.backward(loss.dinputs)
dvalues2 = activation.dinputs

print ( 'Gradients: combined loss and activation:' )
print (dvalues1)
print ( 'Gradients: separate loss and activation:' )
print (dvalues2)


Gradients: combined loss and activation:
[[-0.1         0.03333333  0.06666667]
 [ 0.03333333 -0.16666667  0.13333333]
 [ 0.00666667 -0.03333333  0.02666667]]
Gradients: separate loss and activation:
[[-0.09999999  0.03333334  0.06666667]
 [ 0.03333334 -0.16666667  0.13333334]
 [ 0.00666667 -0.03333333  0.02666667]]


The results are the same.

The small difference between values in both arrays results from the precision of floating-point values in raw Python and NumPy.

Let's compare the execution time of both methods by running both solutions multiple times:

In [27]:
import numpy as np
from timeit import timeit
import nnfs

nnfs.init()

softmax_outputs = np.array([[ 0.7 , 0.1 , 0.2 ],
                            [ 0.1 , 0.5 , 0.4 ],
                            [ 0.02 , 0.9 , 0.08 ]])

class_targets = np.array([ 0 , 1 , 1 ])

def f1 ():
    softmax_loss = Activation_Softmax_Loss_CategoricalCrossentropy()
    softmax_loss.backward(softmax_outputs, class_targets)
    dvalues1 = softmax_loss.dinputs

def f2 ():
    activation = Activation_Softmax()
    activation.output = softmax_outputs
    loss = Loss_CategoricalCrossentropy()
    loss.backward(softmax_outputs, class_targets)
    activation.backward(loss.dinputs)
    dvalues2 = activation.dinputs

t1 = timeit( lambda : f1(), number = 10000 )
t2 = timeit( lambda : f2(), number = 10000 )
print (t2 / t1)

4.83564665840625


Calculating the gradients separately is about 7 times slower, this factor can differ from a machine to a machine.

So, it is worth putting in additional effort to calculate and code the optimized solution of the combined loss and activation function derivative.

## 9.13. Full model code

In [28]:
import numpy as np
import nnfs
from nnfs.datasets import spiral_data

nnfs.init()


# Dense layer
class Layer_Dense:

    # Layer initialization
    def __init__(self, n_inputs, n_neurons):
        # Initialize weights and biases
        self.weights = 0.01 * np.random.randn(n_inputs, n_neurons)
        self.biases = np.zeros((1, n_neurons))

    # Forward pass
    def forward(self, inputs):
        # Remember input values
        self.inputs = inputs
        # Calculate output values from inputs, weights and biases
        self.output = np.dot(inputs, self.weights) + self.biases

    # Backward pass
    def backward(self, dvalues):
        # Gradients on parameters
        self.dweights = np.dot(self.inputs.T, dvalues)
        self.dbiases = np.sum(dvalues, axis=0, keepdims=True)
        # Gradient on values
        self.dinputs = np.dot(dvalues, self.weights.T)


# ReLU activation
class Activation_ReLU:

    # Forward pass
    def forward(self, inputs):

        # Remember input values
        self.inputs = inputs
        # Calculate output values from inputs
        self.output = np.maximum(0, inputs)

    # Backward pass
    def backward(self, dvalues):
        # Since we need to modify original variable,
        # let's make a copy of values first
        self.dinputs = dvalues.copy()

        # Zero gradient where input values were negative
        self.dinputs[self.inputs <= 0] = 0


# Softmax activation
class Activation_Softmax:

    # Forward pass
    def forward(self, inputs):
        # Remember input values
        self.inputs = inputs

        # Get unnormalized probabilities
        exp_values = np.exp(inputs - np.max(inputs, axis=1,
                                            keepdims=True))
        # Normalize them for each sample
        probabilities = exp_values / np.sum(exp_values, axis=1,
                                            keepdims=True)

        self.output = probabilities

    # Backward pass
    def backward(self, dvalues):

        # Create uninitialized array
        self.dinputs = np.empty_like(dvalues)

        # Enumerate outputs and gradients
        for index, (single_output, single_dvalues) in \
                enumerate(zip(self.output, dvalues)):
            # Flatten output array
            single_output = single_output.reshape(-1, 1)
            # Calculate Jacobian matrix of the output
            jacobian_matrix = np.diagflat(single_output) - \
                              np.dot(single_output, single_output.T)

            # Calculate sample-wise gradient
            # and add it to the array of sample gradients
            self.dinputs[index] = np.dot(jacobian_matrix,
                                         single_dvalues)


# Common loss class
class Loss:

    # Calculates the data and regularization losses
    # given model output and ground truth values
    def calculate(self, output, y):

        # Calculate sample losses
        sample_losses = self.forward(output, y)

        # Calculate mean loss
        data_loss = np.mean(sample_losses)

        # Return loss
        return data_loss


# Cross-entropy loss
class Loss_CategoricalCrossentropy(Loss):

    # Forward pass
    def forward(self, y_pred, y_true):

        # Number of samples in a batch
        samples = len(y_pred)

        # Clip data to prevent division by 0
        # Clip both sides to not drag mean towards any value
        y_pred_clipped = np.clip(y_pred, 1e-7, 1 - 1e-7)

        # Probabilities for target values -
        # only if categorical labels
        if len(y_true.shape) == 1:
            correct_confidences = y_pred_clipped[
                range(samples),
                y_true
            ]


        # Mask values - only for one-hot encoded labels
        elif len(y_true.shape) == 2:
            correct_confidences = np.sum(
                y_pred_clipped * y_true,
                axis=1
            )

        # Losses
        negative_log_likelihoods = -np.log(correct_confidences)
        return negative_log_likelihoods

    # Backward pass
    def backward(self, dvalues, y_true):

        # Number of samples
        samples = len(dvalues)
        # Number of labels in every sample
        # We'll use the first sample to count them
        labels = len(dvalues[0])

        # If labels are sparse, turn them into one-hot vector
        if len(y_true.shape) == 1:
            y_true = np.eye(labels)[y_true]

        # Calculate gradient
        self.dinputs = -y_true / dvalues
        # Normalize gradient
        self.dinputs = self.dinputs / samples


# Softmax classifier - combined Softmax activation
# and cross-entropy loss for faster backward step
class Activation_Softmax_Loss_CategoricalCrossentropy():

    # Creates activation and loss function objects
    def __init__(self):
        self.activation = Activation_Softmax()
        self.loss = Loss_CategoricalCrossentropy()

    # Forward pass
    def forward(self, inputs, y_true):
        # Output layer's activation function
        self.activation.forward(inputs)
        # Set the output
        self.output = self.activation.output
        # Calculate and return loss value
        return self.loss.calculate(self.output, y_true)


    # Backward pass
    def backward(self, dvalues, y_true):

        # Number of samples
        samples = len(dvalues)

        # If labels are one-hot encoded,
        # turn them into discrete values
        if len(y_true.shape) == 2:
            y_true = np.argmax(y_true, axis=1)

        # Copy so we can safely modify
        self.dinputs = dvalues.copy()
        # Calculate gradient
        self.dinputs[range(samples), y_true] -= 1
        # Normalize gradient
        self.dinputs = self.dinputs / samples


# Create dataset
X, y = spiral_data(samples=100, classes=3)

# Create Dense layer with 2 input features and 3 output values
dense1 = Layer_Dense(2, 3)

# Create ReLU activation (to be used with Dense layer):
activation1 = Activation_ReLU()

# Create second Dense layer with 3 input features (as we take output
# of previous layer here) and 3 output values (output values)
dense2 = Layer_Dense(3, 3)

# Create Softmax classifier's combined loss and activation
loss_activation = Activation_Softmax_Loss_CategoricalCrossentropy()

# Perform a forward pass of our training data through this layer
dense1.forward(X)

# Perform a forward pass through activation function
# takes the output of first dense layer here
activation1.forward(dense1.output)

# Perform a forward pass through second Dense layer
# takes outputs of activation function of first layer as inputs
dense2.forward(activation1.output)

# Perform a forward pass through the activation/loss function
# takes the output of second dense layer here and returns loss
loss = loss_activation.forward(dense2.output, y)
# Let's see output of the first few samples:
print(loss_activation.output[:5])

# Print loss value
print('loss:', loss)

# Calculate accuracy from output of activation2 and targets
# calculate values along first axis
predictions = np.argmax(loss_activation.output, axis=1)
if len(y.shape) == 2:
    y = np.argmax(y, axis=1)
accuracy = np.mean(predictions==y)

# Print accuracy
print('acc:', accuracy)

# Backward pass
loss_activation.backward(loss_activation.output, y)
dense2.backward(loss_activation.dinputs)
activation1.backward(dense2.dinputs)
dense1.backward(activation1.dinputs)

# Print gradients
print(dense1.dweights)
print(dense1.dbiases)
print(dense2.dweights)
print(dense2.dbiases)

[[0.33333334 0.33333334 0.33333334]
 [0.3333332  0.3333332  0.33333364]
 [0.3333329  0.33333293 0.3333342 ]
 [0.3333326  0.33333263 0.33333477]
 [0.33333233 0.3333324  0.33333528]]
loss: 1.0986104
acc: 0.34
[[ 1.5766357e-04  7.8368583e-05  4.7324400e-05]
 [ 1.8161038e-04  1.1045573e-05 -3.3096312e-05]]
[[-3.60553473e-04  9.66117223e-05 -1.03671395e-04]]
[[ 5.44109462e-05  1.07411419e-04 -1.61822361e-04]
 [-4.07913431e-05 -7.16780924e-05  1.12469446e-04]
 [-5.30112993e-05  8.58172934e-05 -3.28059905e-05]]
[[-1.0729185e-05 -9.4610732e-06  2.0027859e-05]]
